pkg/metallb/l2advertisement.go

```diff
@@ -0,0 +1,416 @@
+
+const (
+	l2AdvertisementKind = "L2Advertisement"
+)
+
+// L2AdvertisementBuilder provides struct for the L2Advertisement object containing connection to
+// the cluster and the L2Advertisement definitions.
+type L2AdvertisementBuilder struct {
+	Definition *mlbtypes.L2Advertisement
+	Object     *mlbtypes.L2Advertisement
+	apiClient  *clients.Settings
+	errorMsg   string
+}
+
+// L2AdvertisementAdditionalOptions additional options for L2Advertisement object.
+type L2AdvertisementAdditionalOptions func(builder *L2AdvertisementBuilder) (*L2AdvertisementBuilder, error)
+
+// NewL2AdvertisementBuilder creates a new instance of L2AdvertisementBuilder.
+func NewL2AdvertisementBuilder(apiClient *clients.Settings, name, nsname string) *L2AdvertisementBuilder {
+	glog.V(100).Infof(
+		"Initializing new L2Advertisement structure with the following params: %s, %s",
+		name, nsname)
+
+	builder := L2AdvertisementBuilder{
+		apiClient: apiClient,
+		Definition: &mlbtypes.L2Advertisement{
+			TypeMeta: metaV1.TypeMeta{
+				Kind:       l2AdvertisementKind,
+				APIVersion: fmt.Sprintf("%s/%s", APIGroup, APIVersion),
+			},
+			ObjectMeta: metaV1.ObjectMeta{
+				Name:      name,
+				Namespace: nsname,
+			}, Spec: mlbtypes.L2AdvertisementSpec{},
+		},
+	}
+
+	if name == "" {
+		glog.V(100).Infof("The name of the L2Advertisement is empty")
+
+		builder.errorMsg = "L2Advertisement 'name' cannot be empty"
+	}
+
+	if nsname == "" {
+		glog.V(100).Infof("The namespace of the L2Advertisement is empty")
+
+		builder.errorMsg = "L2Advertisement 'nsname' cannot be empty"
+	}
+
+	return &builder
+}
+
+// Exists checks whether the given L2Advertisement exists.
+func (builder *L2AdvertisementBuilder) Exists() bool {
+	if valid, _ := builder.validate(); !valid {
+		return false
+	}
+
+	glog.V(100).Infof(
+		"Checking if L2Advertisement %s exists in namespace %s",
+		builder.Definition.Name, builder.Definition.Namespace)
+
+	var err error
+	builder.Object, err = builder.Get()
+
+	return err == nil || !k8serrors.IsNotFound(err)
+}
+
+// Get returns L2Advertisement object if found.
+func (builder *L2AdvertisementBuilder) Get() (*mlbtypes.L2Advertisement, error) {
+	if valid, err := builder.validate(); !valid {
+		return nil, err
+	}
+
+	glog.V(100).Infof(
+		"Collecting L2Advertisement object %s in namespace %s",
+		builder.Definition.Name, builder.Definition.Namespace)
+
+	unsObject, err := builder.apiClient.Resource(
+		GetL2AdvertisementGVR()).Namespace(builder.Definition.Namespace).Get(
+		context.TODO(), builder.Definition.Name, metaV1.GetOptions{})
+
+	if err != nil {
+		glog.V(100).Infof(
+			"L2Advertisement object %s doesn't exist in namespace %s",
+			builder.Definition.Name, builder.Definition.Namespace)
+
+		return nil, err
+	}
+
+	return builder.convertToStructured(unsObject)
+}
+
+// PullL2Advertisement pulls existing L2Advertisement from cluster.
+func PullL2Advertisement(apiClient *clients.Settings, name, nsname string) (*L2AdvertisementBuilder, error) {
```

Comment:
Can you please move this function right after NewL2AdvertisementBuilder?

Reply:
moved